src/pke/lib/scheme/ckksrns/ckksrns-parametergeneration.cpp

```diff
@@ -125,107 +131,122 @@ bool ParameterGenerationCKKSRNS::ParamsGenCKKSRNS(std::shared_ptr<CryptoParamete
                         qPrev      = PreviousPrime(qPrev, cyclOrder);
-                        for (uint32_t j = i + 1; j < numPrimes; j++) {
+                        for (size_t j = i + 1; j < numPrimes; j++) {
                             if (qPrev == moduliQ[j]) {
                                 hasSameMod = true;
+                                break;
                             }
                         }
                     }
                     moduliQ[i] = qPrev;
                 }
                 else {
-                    NativeInteger sfInt = std::llround(sf);
-                    NativeInteger sfRem = sfInt.Mod(cyclOrder);
                     NativeInteger qNext = sfInt + NativeInteger(cyclOrder) - sfRem + NativeInteger(1);
-                    bool hasSameMod     = true;
                     while (hasSameMod) {
                         hasSameMod = false;
                         qNext      = NextPrime(qNext, cyclOrder);
-                        for (uint32_t j = i + 1; j < numPrimes; j++) {
+                        for (size_t j = i + 1; j < numPrimes; j++) {
                             if (qNext == moduliQ[j]) {
                                 hasSameMod = true;
+                                break;
                             }
                         }
                     }
                     moduliQ[i] = qNext;
                 }
+                if (moduliQ[i] > maxPrime)
+                    maxPrime = moduliQ[i];
+                else if (moduliQ[i] < minPrime)
+                    minPrime = moduliQ[i];
 
                 rootsQ[i] = RootOfUnity(cyclOrder, moduliQ[i]);
-                cnt++;
             }
         }
     }
 
     if (firstModSize == dcrtBits) {  // this requires dcrtBits < 60
-        moduliQ[0] = PreviousPrime<NativeInteger>(qPrev, cyclOrder);
+        moduliQ[0] = NextPrime<NativeInteger>(maxPrime, cyclOrder);
     }
     else {
         moduliQ[0] = LastPrime<NativeInteger>(firstModSize, cyclOrder);
+
+        // find if the value of moduliQ[0] is already in the vector starting with moduliQ[1] and
+        // if there is, then get another prime for moduliQ[0]
+        const auto pos = std::find(moduliQ.begin() + 1, moduliQ.end(), moduliQ[0]);
+        if (pos != moduliQ.end()) {
+            moduliQ[0] = NextPrime<NativeInteger>(maxPrime, cyclOrder);
+        }
     }
+    if (moduliQ[0] > maxPrime)
+        maxPrime = moduliQ[0];
+
     rootsQ[0] = RootOfUnity(cyclOrder, moduliQ[0]);
 
     if (scalTech == FLEXIBLEAUTOEXT) {
+        // moduliQ[numPrimes] must still be 0, so it has to be populated now
+
         // no need for extra checking as extraModSize is automatically chosen by the library
-        moduliQ[numPrimes] = FirstPrime<NativeInteger>(extraModSize - 1, cyclOrder);
-        rootsQ[numPrimes]  = RootOfUnity(cyclOrder, moduliQ[numPrimes]);
+        auto tempMod = FirstPrime<NativeInteger>(extraModSize - 1, cyclOrder);
+        if (tempMod > maxPrime)
```

Comment:
Why is this code needed? tempMod can be larger than maxPrime only if it is not in moduliQ. Yet, you use maxPrime (below) only when tempMod is in moduliQ. So the branch and re-assignment of maxtPrime to tempMod is the code that is never executed.